lib/devices/android/android-hybrid.js

```diff
@@ -118,101 +119,129 @@ androidHybrid.getProcessNameFromWebview = function (webview, cb) {
+      caps.chromeOptions[option] = val;
+    });
+  }
+  caps = this.decorateChromeOptions(caps);
+  // see note in chrome.js::createSession for explanation of this pattern
+  this.chromedriver.once('error', ocb);
+  this.chromedriver.on('stateChanged', function (msg) {
+    if (msg.state === 'online') {
+      // save the chromedriver object under the context
       this.sessionChromedrivers[context] = this.chromedriver;
+      // let whoever called us know that we're done setting up session
       ocb();
-    }.bind(this));
-  }.bind(this);
+    } else if (msg.state === 'stopped') {
+      // bind our stop/exit handler, passing in context so we know which
+      // one stopped
+      this.onChromedriverStop(context);
+    }
+  }.bind(this));
+  this.chromedriver.start(caps);
+};
 
-  if (this.sessionChromedrivers[context]) {
-    logger.debug("Found existing Chromedriver for context '" + context + "'." +
-                 " Using it.");
-    this.rememberProxyState();
-    this.chromedriver = this.sessionChromedrivers[context];
-    this.proxyHost = this.chromedriver.proxyHost;
-    this.proxyPort = this.chromedriver.proxyPort;
-    this.proxySessionId = this.chromedriver.chromeSessionId;
-    this.isProxy = true;
 
-    // check the status by sending a simple window-based command to ChromeDriver
-    // if there is an error, we want to recreate the ChromeDriver session
-    var url = '/wd/hub/session/' + this.proxySessionId + '/url';
-    this.chromedriver.proxyTo(url, 'GET', {}, function (err, res, body) {
-      body = JSON.parse(body);
-      if ((body.status === status.codes.NoSuchWindow.code && body.value.message.indexOf('no such window: window was already closed') !== -1) ||
-          (body.status === 100 && body.value.message.indexOf('chrome not reachable') !== -1)) {
-        // the window is closed and we need to reinitialize
-        logger.debug("ChromeDriver is not associated with a window. Re-initializing the session.");
-        this.cleanupChromedriver(this.chromedriver, function () {
-          setupNewChromeDriver(cb);
-        });
-      } else {
+androidHybrid.setupExistingChromedriver = function (context, cb) {
+  logger.debug("Found existing Chromedriver for context '" + context + "'." +
+               " Using it.");
+  this.rememberProxyState();
+  this.chromedriver = this.sessionChromedrivers[context];
+  this.proxyReqRes = this.chromedriver.proxyReq.bind(this.chromedriver);
+  this.isProxy = true;
+
+  // check the status by sending a simple window-based command to ChromeDriver
+  // if there is an error, we want to recreate the ChromeDriver session
+  this.chromedriver.hasWorkingWebview().then(function (works) {
+    if (works) return cb();
+    logger.debug("ChromeDriver is not associated with a window. " +
+                 "Re-initializing the session.");
+    // catch any errors the restart process bubbles up
+    this.chromedriver.once('error', cb);
+    this.chromedriver.on('stateChanged', function (msg) {
```

Comment:
First param is msg not error. I think here we need to move the promise bit we use in test to appium-support, or use a cb equivalent.